crates/router/src/core/payments/routing.rs

```
@@ -1279,34 +1282,83 @@ pub fn make_dsl_input_for_surcharge(
+        .transpose()
+        .change_context(errors::RoutingError::DeserializationError {
+            from: "JSON".to_string(),
+            to: "DynamicRoutingAlgorithmRef".to_string(),
+        })
+        .attach_printable("unable to deserialize DynamicRoutingAlgorithmRef from JSON")?
+        .ok_or(errors::RoutingError::GenericNotFoundError {
+            field: "dynamic_routing_algorithm".to_string(),
+        })?;
+
+    let mut connector_list = routable_connectors.clone();
 
-    let success_based_algo_ref = success_based_dynamic_routing_algo_ref
+    connector_list = dynamic_routing_algo_ref
         .success_based_algorithm
-        .ok_or(errors::RoutingError::GenericNotFoundError { field: "success_based_algorithm".to_string() })
-        .attach_printable(
-            "success_based_algorithm not found in dynamic_routing_algorithm from business_profile table",
-        )?;
+        .as_ref()
+        .async_map(|algorithm| {
+            perform_success_based_routing(
+                state,
+                routable_connectors.clone(),
+                profile,
+                dynamic_routing_config_params_interpolator.clone(),
+                algorithm.clone(),
+            )
+        })
+        .await
+        .transpose()
+        .attach_printable("unable to perform success_based_routing")
+        .ok()
+        .flatten()
+        .unwrap_or(routable_connectors.clone());
+
+    connector_list = dynamic_routing_algo_ref
+        .contract_based_routing
+        .as_ref()
+        .async_map(|algorithm| {
+            perform_contract_based_routing(
+                state,
+                routable_connectors.clone(),
+                profile,
+                dynamic_routing_config_params_interpolator,
+                algorithm.clone(),
+            )
+        })
+        .await
+        .transpose()
+        .attach_printable("unable to perform contract_based_routing")
+        .ok()
+        .flatten()
```

why are you ignoring error here?